pkg/policy/mapstate.go

```diff
@@ -435,12 +443,20 @@ func (e *MapStateEntry) Merge(entry *MapStateEntry) {
 	// Deny entries have no proxy redirection nor auth requirement
 	if e.IsDeny {
 		e.ProxyPort = 0
+		e.Listener = ""
+		e.priority = 0
 		e.hasAuthType = DefaultAuthType
 		e.AuthType = AuthTypeDisabled
 	} else {
 		// Proxy port takes precedence, but may be updated
-		if entry.ProxyPort != 0 {
-			e.ProxyPort = entry.ProxyPort
+		if entry.IsRedirectEntry() {
+			// Lower number has higher priority
+			if !e.IsRedirectEntry() ||
+				(entry.ProxyPort != e.ProxyPort && entry.priority < e.priority) {
```

Comment:
I'm not sure I'm correctly understanding the logic here. Shouldn't we overwrite `e.Listener` and `e.Priority` even when `entry.ProxyPort == e.ProxyPort`?

Reply:
Each listener has a unique proxy port, so when the proxy ports are the same, the listeners should already have the same value. But this brings up the point that now that MapStateEntry contains the correct proxy port from the beginning (3rd commit), we might get away with without the listener name in the MapSteteEntry (added in the 2nd commit, as the proxy port was not initially valid). I'll check this possibility in a follow-up.

But you are absolutely correct about the priority, as it is not an intrinsic property of the listener itself, but of the reference to a listener. Different references to the same listener could have different priorities, which needs to be properly propagated. I'll fix this immediately. Turns out that the simplest form is what you were after, overwriting also values that may already be the same:
```
		// Proxy port takes precedence, but may be updated due to priority
		if entry.IsRedirectEntry() {
			// Lower number has higher priority, but non-redirects have 0 priority value
			if !e.IsRedirectEntry() || entry.priority < e.priority {
				e.ProxyPort = entry.ProxyPort
				e.Listener = entry.Listener
				e.priority = entry.priority
			}
		}
```
While adding unit testing for the above I found two additional problems:
1. I had missed updating the newish MapStateEntry.DeepEqual function when rebasing this to main (DeepEqual was added in #26344)
2. mapState.addKeyWithChanges() has a bug where DatapathEqual() is evaluated after the entries are already merged (so they are datapath-equal in all cases). This bug was introduced in #26331, which has not been released yet, so no backports needed. Fixed in a separate commit, will note in PR description.

Reply:
Thanks for the explanation :pray: